lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -991,26 +1002,136 @@ class PullbackCloner::Implementation final
+    // bb8(%39 : @owned $@callee_guaranteed (Float) -> Float)
+    //   %40 = apply %39(%0) : $@callee_guaranteed (Float) -> Float
+    //   accumulate adjoints (using %35)...
+    //   %54 = load [trivial] %42
+    //   br bb10(%54)
+    // bb9:
+    //   br bb10(%35)
+    // bb10(%60 : $Float):
+    //   ...
+    auto loc = tai->getLoc();
+    auto optPullback = getPullbackTupleElement(tai);
+    auto *pbBB = builder.getInsertionBB();
+    auto *normalPbBB = pbBB->split(builder.getInsertionPoint());
+    auto *errorPbBB = getPullback().createBasicBlockAfter(normalPbBB);
+    auto *afterTryApplyPbBB = getPullback().createBasicBlockAfter(errorPbBB);
+
+    // Note that we cannot simply assign map value as DenseMap entry references
+    // could be invalidated on insertion, so map[a] = map[b] may trigger
+    // use-after-free
+    auto currentBT = blockTemporaries[pbBB];
+    blockTemporaries.insert({afterTryApplyPbBB, currentBT});
+    blockTemporaries[pbBB].clear();
+
+    auto pullback =
+      builder.createSwitchOptional(loc, optPullback,
+                                   normalPbBB, errorPbBB,
+                                   optPullback->getOwnershipKind());
+
+    SmallVector<SILValue, 2> adjArgs;
+    FullApplySite fai(tai);
+    auto *originalBB = tai->getParent();
     auto &applyInfo = applyInfoLookup->getSecond();
+    for (unsigned i : applyInfo.config.parameterIndices->getIndices()) {
+      unsigned argIdx = fai.getNumIndirectSILResults() + fai.getNumIndirectSILErrorResults() + i;
+      auto origArg = fai.getArgument(argIdx);
+      auto paramInfo = fai.getSubstCalleeConv().getParamInfoForSILArg(argIdx);
+      if (paramInfo.isAutoDiffSemanticResult())
+        continue;
+      if (getTangentValueCategory(origArg) != SILValueCategory::Object)
+        continue;
+
+      adjArgs.push_back(origArg);
+      afterTryApplyPbBB->createPhiArgument(origArg->getType(),
+                                           OwnershipKind::Owned);
+    }
+
+    {
+      builder.setInsertionPoint(errorPbBB);
+
+      SmallVector<SILValue> outAdjArgs;
+      for (auto arg : adjArgs) {
+        auto argAdj = getAdjointValue(originalBB, arg);
+        outAdjArgs.push_back(materializeAdjointDirect(argAdj, loc));
+      }
+
+      cleanUpTemporariesForBlock(errorPbBB, loc);
+      builder.createBranch(loc, afterTryApplyPbBB, outAdjArgs);
+    }
+
+    {
+      builder.setInsertionPoint(normalPbBB);
+      buildPullbackCall(tai, applyInfo, pullback);
+
+      SmallVector<SILValue> outAdjArgs;
+      for (auto arg : adjArgs) {
+        auto argAdj = getAdjointValue(originalBB, arg);
+        outAdjArgs.push_back(materializeAdjointDirect(argAdj, loc));
+      }
+
+      cleanUpTemporariesForBlock(normalPbBB, loc);
+      builder.createBranch(loc, afterTryApplyPbBB, outAdjArgs);
+    }
+
+    builder.setInsertionPoint(afterTryApplyPbBB);
+    for (auto argAndIdx : llvm::enumerate(adjArgs)) {
```

Comment:
Nit: probably worth using structured bindings and replacing `.value()` and `.index()` below with corresponding variables
```suggestion
    for (auto [index, arg] : llvm::enumerate(adjArgs)) {
```